### ISSUE TITLE
Lambda Concatenate behavior changes based on initialization

### ISSUE THREAD

**Issue by nishbo** [keras-team/keras#15914]
**System information**.
- Have I written custom code (as opposed to using a stock example script provided in Keras): Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Microsoft Windows [Version 10.0.19043.1165]
- TensorFlow installed from (source or binary): Using pip
- TensorFlow version (use command below): v2.7.0-rc1-69-gc256c071bb2 2.7.0
- Python version: 3.7.5
- GPU model and memory: NVIDIA GeForce RTX 2060 6GBytes

**The problem**.

Hi, I am investigating how constraints on the data separation within layers alter model behavior, and to this end I create functional models that separately process slices of input. In the simplest case, I have an array of indices `ids=list(range(8))` within input data tensor, I make Lambda layers for each index, Concatenate them, and then feed into a Dense layer. I have found that depending on how I create Lambda layers I get drastically different results:

The first network:
```
input_layer = tf.keras.Input(shape=(input_dim,))

lambdas = [
    tf.keras.layers.Lambda(lambda x: x[:, pi:pi+1], output_shape=((1,)))(input_layer)
    for pi in ids]
input_section = tf.keras.layers.Concatenate()(lambdas)

layer = tf.keras.layers.Dense(n_nodes, activation='relu')(input_section)
out_layer = tf.keras.layers.Dense(n_classes, activation='sigmoid')(layer)
model = tf.keras.models.Model(input_layer, out_layer)
```
The second network:
```
input_layer = tf.keras.Input(shape=(input_dim,))

lambda0 = tf.keras.layers.Lambda(lambda x: x[:, ids[0]:ids[0]+1], output_shape=((1,)))(input_layer)
lambda1 = tf.keras.layers.Lambda(lambda x: x[:, ids[1]:ids[1]+1], output_shape=((1,)))(input_layer)
lambda2 = tf.keras.layers.Lambda(lambda x: x[:, ids[2]:ids[2]+1], output_shape=((1,)))(input_layer)
lambda3 = tf.keras.layers.Lambda(lambda x: x[:, ids[3]:ids[3]+1], output_shape=((1,)))(input_layer)
lambda4 = tf.keras.layers.Lambda(lambda x: x[:, ids[4]:ids[4]+1], output_shape=((1,)))(input_layer)
lambda5 = tf.keras.layers.Lambda(lambda x: x[:, ids[5]:ids[5]+1], output_shape=((1,)))(input_layer)
lambda6 = tf.keras.layers.Lambda(lambda x: x[:, ids[6]:ids[6]+1], output_shape=((1,)))(input_layer)
lambda7 = tf.keras.layers.Lambda(lambda x: x[:, ids[7]:ids[7]+1], output_shape=((1,)))(input_layer)
input_section = tf.keras.layers.Concatenate()((lambda0, lambda1, lambda2, lambda3, lambda4, lambda5, lambda6, lambda7))

layer = tf.keras.layers.Dense(n_nodes, activation='relu')(input_section)
out_layer = tf.keras.layers.Dense(n_classes, activation='sigmoid')(layer)
model = tf.keras.models.Model(input_layer, out_layer)
```

The first network always produces classification quality on average around 5%, and the second one - 20%. I got larger differences when ids would span the whole length of input layer (up to 40%). I noticed that if I increase the slice length in the first model to 8 (effectively making the first Lambda to be equivalent to the whole Concatenated layer), the model performance rises to the appropriate level (20%). Similarly, if in the second model, I run Concatenate only on lambda0, it peaks at 5% accuracy.

The networks seem the same, I get exactly the same figures when I plot them using `plot_model` or inspect their summaries (attached in the end)

I am currently trying to go through `tensorflow/python/keras/layers/core.py@class Lambda`, but do not see any obvious differences in the created Lambda objects yet.

**Source code / logs**.
Models are compiled and fit with the same parameters and run for the same number of epochs.

Model summaries:
```Model: "model"
__________________________________________________________________________________________________
 Layer (type)                   Output Shape         Param #     Connected to
==================================================================================================
 input_1 (InputLayer)           [(None, 36)]         0           []

 lambda (Lambda)                (None, 1)            0           ['input_1[0][0]']

 lambda_1 (Lambda)              (None, 1)            0           ['input_1[0][0]']

 lambda_2 (Lambda)              (None, 1)            0           ['input_1[0][0]']

 lambda_3 (Lambda)              (None, 1)            0           ['input_1[0][0]']

 lambda_4 (Lambda)              (None, 1)            0           ['input_1[0][0]']

 lambda_5 (Lambda)              (None, 1)            0           ['input_1[0][0]']

 lambda_6 (Lambda)              (None, 1)            0           ['input_1[0][0]']

 lambda_7 (Lambda)              (None, 1)            0           ['input_1[0][0]']

 tf.concat (TFOpLambda)         (None, 8)            0           ['lambda[0][0]',
                                                                  'lambda_1[0][0]',
                                                                  'lambda_2[0][0]',
                                                                  'lambda_3[0][0]',
                                                                  'lambda_4[0][0]',
                                                                  'lambda_5[0][0]',
                                                                  'lambda_6[0][0]',
                                                                  'lambda_7[0][0]']

 dense (Dense)                  (None, 80)           720         ['tf.concat[0][0]']

 dense_1 (Dense)                (None, 32)           2592        ['dense[0][0]']

==================================================================================================
Total params: 3,312
Trainable params: 3,312
Non-trainable params: 0
__________________________________________________________________________________________________

Model: "model_1"
__________________________________________________________________________________________________
 Layer (type)                   Output Shape         Param #     Connected to
==================================================================================================
 input_2 (InputLayer)           [(None, 36)]         0           []

 lambda_8 (Lambda)              (None, 1)            0           ['input_2[0][0]']

 lambda_9 (Lambda)              (None, 1)            0           ['input_2[0][0]']

 lambda_10 (Lambda)             (None, 1)            0           ['input_2[0][0]']

 lambda_11 (Lambda)             (None, 1)            0           ['input_2[0][0]']

 lambda_12 (Lambda)             (None, 1)            0           ['input_2[0][0]']

 lambda_13 (Lambda)             (None, 1)            0           ['input_2[0][0]']

 lambda_14 (Lambda)             (None, 1)            0           ['input_2[0][0]']

 lambda_15 (Lambda)             (None, 1)            0           ['input_2[0][0]']

 tf.concat_1 (TFOpLambda)       (None, 8)            0           ['lambda_8[0][0]',
                                                                  'lambda_9[0][0]',
                                                                  'lambda_10[0][0]',
                                                                  'lambda_11[0][0]',
                                                                  'lambda_12[0][0]',
                                                                  'lambda_13[0][0]',
                                                                  'lambda_14[0][0]',
                                                                  'lambda_15[0][0]']

 dense_2 (Dense)                (None, 80)           720         ['tf.concat_1[0][0]']

 dense_3 (Dense)                (None, 32)           2592        ['dense_2[0][0]']

==================================================================================================
Total params: 3,312
Trainable params: 3,312
Non-trainable params: 0
__________________________________________________________________________________________________
```

**Comment by nishbo** [comment]
I have made a small example demonstrating this problem:
```
import itertools

import numpy as np

import matplotlib
import matplotlib.pyplot as plt

import sklearn
import sklearn.preprocessing

import tensorflow as tf


def get_list_lambda_network(input_dim, n_nodes, n_classes):
    input_layer = tf.keras.Input(shape=(input_dim,))

    lambdas = [
        tf.keras.layers.Lambda(lambda x: x[:, i:i+1], output_shape=((1,)))(input_layer)
        for i in range(input_dim)]
    input_section = tf.keras.layers.Concatenate()(lambdas)

    layer = tf.keras.layers.Dense(n_nodes, activation='relu')(input_section)
    out_layer = tf.keras.layers.Dense(n_classes, activation='sigmoid')(layer)
    model = tf.keras.models.Model(input_layer, out_layer)
    model._name = 'list_lambda_model'
    return model


def get_hardcoded_lambda_network(input_dim, n_nodes, n_classes):
    input_layer = tf.keras.Input(shape=(input_dim,))

    lambda0 = tf.keras.layers.Lambda(lambda x: x[:, 0:1], output_shape=((1,)))(input_layer)
    lambda1 = tf.keras.layers.Lambda(lambda x: x[:, 1:2], output_shape=((1,)))(input_layer)
    input_section = tf.keras.layers.Concatenate()((lambda0, lambda1))

    layer = tf.keras.layers.Dense(n_nodes, activation='relu')(input_section)
    out_layer = tf.keras.layers.Dense(n_classes, activation='sigmoid')(layer)
    model = tf.keras.models.Model(input_layer, out_layer)
    model._name = 'hardcoded_lambda_model'
    return model


def main():
    # two predicting variables vary from 0 to n_x. y is defined as x_1*10 + x2. Class depends on
    # both inputs
    n_x = 2
    n_classes = n_x**2
    # number of times each class is added to the dataset
    ppx = 10
    # normal noise added to x
    noise_sd = 0.05
    n_epochs = 400
    n_nodes = 4*n_classes

    # generate dataset
    rng = np.random.default_rng()
    noise = lambda v: rng.normal(loc=v, scale=noise_sd)
    x = []
    y = []
    for x1, x2 in itertools.product(range(n_x), range(n_x)):
        for _ in range(ppx):
            x.append([noise(x1), noise(x2)])
            y.append(x1*10 + x2)

    # transform to numpy arrays
    x = np.array(x)
    y = np.array(y).reshape(-1, 1)

    # preprocessing
    sc = sklearn.preprocessing.StandardScaler()
    X = sc.fit_transform(x)
    ohe = sklearn.preprocessing.OneHotEncoder()
    y = ohe.fit_transform(y).toarray()

    # test the list lambda model
    list_lambda_model = get_list_lambda_network(n_x, n_nodes, n_classes)
    list_lambda_model.compile(
        loss='categorical_crossentropy',
        optimizer=tf.keras.optimizers.Adam(),
        metrics=['categorical_accuracy'])
    tf.keras.utils.plot_model(
        list_lambda_model, to_file='list_lambda_model.png', show_shapes=True, show_dtype=True,
        show_layer_names=True, show_layer_activations=True)

    list_lambda_history = list_lambda_model.fit(
        X, y,
        epochs=n_epochs,
        use_multiprocessing=True
    )

    # test the hardcoded lambda model
    hardcoded_lambda_model = get_hardcoded_lambda_network(n_x, n_nodes, n_classes)
    hardcoded_lambda_model.compile(
        loss='categorical_crossentropy',
        optimizer=tf.keras.optimizers.Adam(),
        metrics=['categorical_accuracy'])
    tf.keras.utils.plot_model(
        hardcoded_lambda_model, to_file='hardcoded_lambda_model.png', show_shapes=True,
        show_dtype=True, show_layer_names=True, show_layer_activations=True)

    hardcoded_lambda_history = hardcoded_lambda_model.fit(
        X, y,
        epochs=n_epochs,
        use_multiprocessing=True
    )

    print()
    print('List lambda fit: {}. Best performance during training: {}.'.format(
        list_lambda_history.history['categorical_accuracy'][-1],
        max(list_lambda_history.history['categorical_accuracy'])))
    print('Hardcoded lambda fit: {}. Best performance during training: {}.'.format(
        hardcoded_lambda_history.history['categorical_accuracy'][-1],
        max(hardcoded_lambda_history.history['categorical_accuracy'])))

    fig = plt.figure()
    plt.plot(list_lambda_history.epoch, list_lambda_history.history['categorical_accuracy'],
             label='List lambda model')
    plt.plot(hardcoded_lambda_history.epoch, hardcoded_lambda_history.history['categorical_accuracy'],
             label='Hardcoded lambda model')
    plt.xlabel('Epoch #')
    plt.ylabel('Categorical accuracy, nu')
    plt.legend()
    plt.title('Model performance comparison')


if __name__ == '__main__':
    main()

    plt.show()
```

Running with the system described in the first post, I get the following output:
```
 >py lambda_function_bug.py
2022-01-25 11:46:54.675973: I tensorflow/core/platform/cpu_feature_guard.cc:151] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX AVX2
To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.
2022-01-25 11:46:55.152097: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1525] Created device /job:localhost/replica:0/task:0/device:GPU:0 with 3965 MB memory:  -> device: 0, name: NVIDIA GeForce RTX 2060, pci bus id: 0000:01:00.0, compute capability: 7.5
Epoch 1/400
2/2 [==============================] - 1s 3ms/step - loss: 1.4475 - categorical_accuracy: 0.2500
Epoch 2/400
2/2 [==============================] - 0s 3ms/step - loss: 1.4362 - categorical_accuracy: 0.2500
Epoch 3/400
2/2 [==============================] - 0s 3ms/step - loss: 1.4255 - categorical_accuracy: 0.2500
Epoch 4/400
2/2 [==============================] - 0s 3ms/step - loss: 1.4149 - categorical_accuracy: 0.2500
Epoch 5/400
2/2 [==============================] - 0s 3ms/step - loss: 1.4050 - categorical_accuracy: 0.2500
<...>
Epoch 395/400
2/2 [==============================] - 0s 3ms/step - loss: 0.7030 - categorical_accuracy: 0.5000
Epoch 396/400
2/2 [==============================] - 0s 2ms/step - loss: 0.7029 - categorical_accuracy: 0.5000
Epoch 397/400
2/2 [==============================] - 0s 3ms/step - loss: 0.7030 - categorical_accuracy: 0.5000
Epoch 398/400
2/2 [==============================] - 0s 3ms/step - loss: 0.7026 - categorical_accuracy: 0.5000
Epoch 399/400
2/2 [==============================] - 0s 4ms/step - loss: 0.7027 - categorical_accuracy: 0.5000
Epoch 400/400
2/2 [==============================] - 0s 3ms/step - loss: 0.7021 - categorical_accuracy: 0.5000
Epoch 1/400
2/2 [==============================] - 0s 3ms/step - loss: 1.4596 - categorical_accuracy: 0.0000e+00
Epoch 2/400
2/2 [==============================] - 0s 2ms/step - loss: 1.4476 - categorical_accuracy: 0.0500
Epoch 3/400
2/2 [==============================] - 0s 3ms/step - loss: 1.4362 - categorical_accuracy: 0.0750
Epoch 4/400
2/2 [==============================] - 0s 3ms/step - loss: 1.4246 - categorical_accuracy: 0.1000
Epoch 5/400
2/2 [==============================] - 0s 4ms/step - loss: 1.4134 - categorical_accuracy: 0.1250
<..>
Epoch 395/400
2/2 [==============================] - 0s 3ms/step - loss: 0.0324 - categorical_accuracy: 1.0000
Epoch 396/400
2/2 [==============================] - 0s 3ms/step - loss: 0.0322 - categorical_accuracy: 1.0000
Epoch 397/400
2/2 [==============================] - 0s 3ms/step - loss: 0.0320 - categorical_accuracy: 1.0000
Epoch 398/400
2/2 [==============================] - 0s 3ms/step - loss: 0.0318 - categorical_accuracy: 1.0000
Epoch 399/400
2/2 [==============================] - 0s 3ms/step - loss: 0.0316 - categorical_accuracy: 1.0000
Epoch 400/400
2/2 [==============================] - 0s 4ms/step - loss: 0.0314 - categorical_accuracy: 1.0000

List lambda fit: 0.5. Best performance during training: 0.625.
Hardcoded lambda fit: 1.0. Best performance during training: 1.0.
```
Plus images of model architectures and performance:
![Figure_1](https://user-images.githubusercontent.com/3775317/151031532-b5c959db-7f8a-4a9a-a9a2-4ac31e7d8eca.png)

The hardcoded model clearly uses both inputs and is able to achieve 100% performance, while the definition through a list is using only one of the inputs, so it gets 50% of the guesses correctly. Any suggestions are welcome.

**Comment by sushreebarsa** [comment]
@nishbo I tried to replicate the issue on colab using TF v2.7.0 and tf-nightly, could you please find the gist [here](https://colab.research.google.com/gist/sushreebarsa/86b0d7becb90225c9d742e8e7edbe749/15914-keras.ipynb) and confirm the same?Thank you!

**Comment by fchollet** [comment]
Hi,

This is just a standard Python mistake. When you do:

```
lambdas = [lambda x: x[:, pi:pi+1] for pi in ids]
```

what's actually happening is not what you think is happening. You think you're materializing `len(ids)` different lambda functions that each contains the hardcoded value of `pi` at the time of instantiation. What's actually happening is that all your lambda are the *same* and they all refer to `pi`. So when you call them they just use the current value of `pi`, which is the last entry in `ids`.

Try this:

```
ids = [0, 1, 2, 3]
lambdas = [lambda x: print(x[pi]) for pi in ids]

lambdas[0]([0, 1, 2, 3, 4, 5])  # Returns 3
lambdas[1]([0, 1, 2, 3, 4, 5])  # Returns 3
lambdas[2]([0, 1, 2, 3, 4, 5])  # etc.
```

Python is sometimes counter-intuitive. Here's my recommendation: never use `lambda` or `Lambda` layers. Just use your own `Layer` subclasses. Much easier.

**Comment by nishbo** [comment]
@fchollet Thank you, I'll follow your advice.
